### PR TITLE
Improve streamable

### PR DESCRIPTION
**Pull request by richardkiss**

### CODE REVIEW

File: chia/util/streamable.py [hunk 1]
```diff
@@ -152,80 +155,131 @@ def streamable(cls: Any):
...
+        return not_streamable
+
+    for _, f_type in fields.items():
+        parse_functions.append(getattr(cls1, "function_to_parse_one_item", fail)(f_type))
+
+    PARSE_FUNCTIONS_FOR_STREAMABLE_CLASS[t] = parse_functions
+    return t
+
+
+def parse_bool(f):
```
Comment (mariano54) [inline chia/util/streamable.py:179]: All these functions need type annotations

File: chia/util/streamable.py [hunk 2]
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
...
+    for _, f_type in fields.items():
+        parse_functions.append(cls.function_to_parse_one_item(f_type))
+
+    PARSE_FUNCTIONS_FOR_STREAMABLE_CLASS[t] = parse_functions
+    return t
+
+
+def parse_bool(f):
+    bool_byte = f.read(1)
+    assert bool_byte is not None and len(bool_byte) == 1  # Checks for EOF
```
Comment (aqk) [inline chia/util/streamable.py:175]: Will these asserts activate in any run mode, or is it like C, where they can be deactivated in some run modes?

File: chia/util/streamable.py [hunk 3]
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
...
-            return f_type.parse(f)
+            return f_type.parse
         if f_type == bytes:
-            list_size_bytes = f.read(4)
-            assert list_size_bytes is not None and len(list_size_bytes) == 4  # Checks for EOF
-            list_size: uint32 = uint32(int.from_bytes(list_size_bytes, "big"))
-            bytes_read = f.read(list_size)
-            assert bytes_read is not None and len(bytes_read) == list_size
-            return bytes_read
+            return parse_bytes
```
Comment (aqk) [inline chia/util/streamable.py:254]: This section is nice.

File: chia/wallet/wallet_coin_store.py [hunk 4]
```diff
@@ -126,17 +127,23 @@ async def set_spent(self, coin_name: bytes32, height: uint32):
 
         await self.add_coin_record(spent)
 
+    def coin_record_from_row(self, row: sqlite3.Row) -> WalletCoinRecord:
+        coin = Coin(bytes32(bytes.fromhex(row[6])), bytes32(bytes.fromhex(row[5])), uint64.from_bytes(row[7]))
+        return WalletCoinRecord(
+            coin, uint32(row[1]), uint32(row[2]), bool(row[3]), bool(row[4]), WalletType(row[8]), row[9]
```
Comment (aqk) [inline chia/wallet/wallet_coin_store.py:133]: What is the type of row[9] - I notice it is the only one not explicitly cast

File: chia/wallet/wallet_puzzle_store.py [hunk 5]
```diff
@@ -120,11 +120,11 @@ async def get_derivation_record(self, index: uint32, wallet_id: uint32) -> Optio
...
         if row is not None and row[0] is not None:
             return DerivationRecord(
-                row[0],
-                bytes.fromhex(row[2]),
+                uint32(row[0]),
+                bytes32.fromhex(row[2]),
                 G1Element.from_bytes(bytes.fromhex(row[1])),
-                row[3],
-                row[4],
+                WalletType(row[3]),
```
Comment (aqk) [inline chia/wallet/wallet_puzzle_store.py:126]: WalletType: nice.

File: chia/util/streamable.py [hunk 6]
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
...
+    t = type(cls.__name__, (cls1, Streamable), {})
+
+    parse_functions = []
+    try:
+        fields = cls1.__annotations__  # pylint: disable=no-member
+    except Exception:
+        fields = {}
+
+    for _, f_type in fields.items():
+        parse_functions.append(cls.function_to_parse_one_item(f_type))
```
Comment (aqk) [inline chia/util/streamable.py:167]: Is the intent here to automatically add parse functions for only the primitive types that are contained as members?

**Review by aqk — DISMISSED** [review]
Looks good. One thing I don't understand in the first question.

File: chia/util/streamable.py (hunk 6 above)
Comment (richardkiss) [reply chia/util/streamable.py:167]: The general idea is to traverse the `__annotations__` tree just once at class instantiation time, and put into a list the sub-parsing functions which do parsing and not introspection.

File: chia/util/streamable.py (hunk 2 above)
Comment (richardkiss) [reply chia/util/streamable.py:175]: `assert` is removed in `python -OO` mode, which we don't do. So we shouldn't use `assert` for flow control. However, this problem was here before this change, so it doesn't make these semantics any worse.

File: chia/protocols/timelord_protocol.py [hunk 7]
```diff
@@ -74,7 +74,7 @@ class NewEndOfSubSlotVDF(Streamable):
 
 @dataclass(frozen=True)
 @streamable
-class RequestCompactProofOfTime:
+class RequestCompactProofOfTime(Streamable):
```
Comment (arvidn) [inline chia/protocols/timelord_protocol.py:77]: I would expect the annotation could add the base class too, no? (maybe for later)

File: chia/wallet/wallet_coin_store.py (hunk 4 above)
Comment (arvidn) [reply chia/wallet/wallet_coin_store.py:133]: a comment with the schema of this table would be very helpful, but I suppose this is not quite in scope for this PR

File: chia/wallet/wallet_puzzle_store.py [hunk 8]
```diff
@@ -142,11 +142,11 @@ async def get_derivation_record_for_puzzle_hash(self, puzzle_hash: str) -> Optio
 
         if row is not None and row[0] is not None:
             return DerivationRecord(
-                row[0],
-                bytes.fromhex(row[2]),
+                uint32(row[0]),
+                bytes32.fromhex(row[2]),
```
Comment (arvidn) [inline chia/wallet/wallet_puzzle_store.py:146]: we really store binary data in hex form in the database?

File: chia/util/streamable.py [hunk 9]
```diff
@@ -152,80 +155,121 @@ def streamable(cls: Any):
...
         if f_type is str:
-            str_size_bytes = f.read(4)
-            assert str_size_bytes is not None and len(str_size_bytes) == 4  # Checks for EOF
-            str_size: uint32 = uint32(int.from_bytes(str_size_bytes, "big"))
-            str_read_bytes = f.read(str_size)
-            assert str_read_bytes is not None and len(str_read_bytes) == str_size  # Checks for EOF
-            return bytes.decode(str_read_bytes, "utf-8")
-        raise RuntimeError(f"Type {f_type} does not have parse")
+            return parse_str
+        raise NotImplementedError(f"Type {f_type} does not have parse")
```
Comment (arvidn) [inline chia/util/streamable.py:268]: this message doesn't look right to me. Shouldn't it say "Type {f_type} does not have from_bytes()"?

**Review by arvidn — COMMENTED** [review]
I think this approach is a lot nicer than the previous one. It especially makes it easy to test the parser for each primitive type, and really cover all the edge cases. I really think we should do that. This code has better work, everything else relies on it.

I would also have expected to see tests where we ensure that some (or maybe all) network messages encode to specific bytes. Just passing a round-trip test isn't enough to ensure we don't break the protocol.

I'll take a look at adding some tests today.

File: chia/util/streamable.py (hunk 9 above)
Comment (arvidn) [reply chia/util/streamable.py:268]: I see, `parse()` is the main way of supporting the streamable protocol

**Comment by rostislav** [comment]
I performed the following benchmark on old and new Streamable to compare their performance: parsing the first 10000 mainnet blocks. The old result: 1.39s; the new result: 1.08s. The speed improvement is 23%.

File: chia/wallet/wallet_puzzle_store.py (hunk 8 above)
Comment (rostislav) [reply chia/wallet/wallet_puzzle_store.py:146]: Yes, we currently do. See `add_derivation_paths()` in the same file.

File: chia/util/streamable.py (hunk 9 above)
Comment (rostislav) [reply chia/util/streamable.py:268]: Currently `from_bytes()` is used only for sized integer types. All other types that have custom serialization need to provide their own `.parse()` function.

File: chia/protocols/timelord_protocol.py (hunk 7 above)
Comment (rostislav) [reply chia/protocols/timelord_protocol.py:77]: The annotation could do the dataclass wrapping as well, but the comment in `def streamable` states that this is intended for linters so they could recognize the methods of the base class as well as the `__init__` method provided by `dataclass`.

**Comment by mariano54** [comment]
Nice work. If nothing is broken with the wallet, I think we can probable merge it (and if arvids tests pass)

File: chia/util/streamable.py (hunk 1 above)
Comment (arvidn) [reply chia/util/streamable.py:173]: https://github.com/Chia-Network/chia-blockchain/pull/3222

**Review by mariano54 — APPROVED** [review]
I've been running the spend bot with it, no issues
